doc/decisions/0_drafts/keyname.md

```diff
@@ -0,0 +1,136 @@
+
+Internally, this could be combined into one buffer, stored as a single pointer and size, again.
+However, depending on the API there can also be benefits to keeping the namespace as a separate field.
+
+Even with a separate namespace field, the benefits of "Only unescaped name" are retained.
+The memory consumption is near minimal (alignment padding can cause a difference).
+Comparisons are exactly the same, just with an additional namespace byte comparison beforehand.
+
+### Both escaped and unescaped name
+
+To combine the advantages of escaped and unescaped name, both could be used.
+
+The API could largely rely on the escaped name, while e.g., comparisons can use the unescaped name.
+
+The issue with this approach is the insane memory consumption.
+Keynames can already be quite long and `Key` is at the base of Elektra.
+Storing every name twice in only slightly different forms essentially doubles the memory consumption.
+
+### Both escaped and unescaped name, but only unescaped stored
+
+Instead of storing both escaped and unescaped name, only the unescaped name could be stored.
+
+APIs that use the escaped name would do conversion on the fly.
+
+This approach has several downsides.
+First, while the conversion may be optimized, it will never be free in terms of runtime.
+But more importantly, if an escaped name should be returned by an API, it must be stored somewhere.
+This means extra allocations and crucially somebody needs to do the cleanup.
+In other words, it complicates the API.
+
+### Escaped and unescaped name in single buffer
+
+Another variant of the above.
+The escaped and unescaped name are stored in a single buffer.
+This avoids extra allocations and extra pointers and sizes in structs.
+
+The escaped name could also be stored lazily only when needed.
+This would solve the cleanup problem.
+
+While this may seem like the ideal solution, there are still some downsides.
+The biggest problem is the API design.
+If the API uses escaped names a lot (because it is more convenient), then this essentially degrades into the "Both escaped and unescaped name" solution.
+Even if APIs exist for both escaped and unescaped names, the convenience benefit, will lead to more use of escaped names.
+This means the escaped name will be stored for many keys and therefore the benefit of the lazy allocation is negated.
+
+Without the lazy allocation benefit, the only difference to "Both escaped and unescaped name" is that we have fewer pointers and sizes in structs.
+This saves some amount of memory and allocations, but makes internal code more difficult to write and understand.
+
+## Decision
+
+**Suggestion:**
+
+- Go with "Only unescaped name, with separate namespace" from above
+- Leave whether namespace is stored separately TBD, until API requirements are worked out
+
+## Rationale
+
+- Largest memory savings
+- Still useful API
```

Comment:
Proposed API not clear to me.

Reply:
I'll explain it some more soon. The "Decision" and "Rationale" are not very detailed, because I wasn't sure how quick this PR would progress...